field/TinyLang/Field/Generator.hs

```diff
@@ -228,81 +227,126 @@ arbitraryUnOpRing = elements [Neg]
 arbitraryBinOpRing :: MonadGen m => m (BinOp f (AField f) (AField f) (AField f))
 arbitraryBinOpRing = elements [Add, Sub, Mul]
 
-groundArbitraryFreqs
-    :: (Field f, Arbitrary f, KnownUni f a, MonadGen m)
+groundArbitraryFreqs :: (Field f, Arbitrary f, KnownUni f a, MonadGen m)
     => Vars f -> [(Int, m (Expr f a))]
 groundArbitraryFreqs vars =
     [ (1, EConst <$> arbitraryM)
     , (2, EVar   <$> chooseUniVar vars)
     ]
 
--- | Generate an expression of a particular type from a collection of variables
+newtype SGenT f (m :: Type -> Type) a =
+    SGen { unSGenT :: GenT (SupplyT (StateT (Vars f) m)) a }
+    deriving newtype ( Monad
+                     , Functor
+                     , Applicative
+                     , MonadSupply
+                     , MonadState (Vars f)
+                     , MonadGen
+                     )
+
+-- TODO:  Is there a way to automate this?
+instance MonadState s m => MonadState s (GenT m) where
+    get = lift get
+    put = lift . put
+    state = lift . state
+
+
+type SGen f a = SGenT f Identity a
+
+-- Not sure if this is the right type
+runSGenT :: (Monad m) => Vars f -> SGenT f m a -> Gen (m a)
+runSGenT vars = fmap ((flip evalStateT) vars) . fmap (runSupplyT) . runGenT . unSGenT
```

Comment:
Normally, I'd make it

```haskell
runSGenT :: Monad m => Vars f -> SGenT f m a -> GenT m a
```

by using [`hoist`](http://hackage.haskell.org/package/mmorph-1.1.3/docs/Control-Monad-Morph.html#v:hoist), but `GenT` doesn't have an `MFunctor` instance, but you could provide one. Not a big deal.